src/zarr/storage/_cache.py

```diff
@@ -0,0 +1,698 @@
+            self._max_size,
+            self._current_size,
+            self._keys_cache,
+            self._contains_cache,
+            self._listdir_cache,
+            self._values_cache,
+            self.hits,
+            self.misses,
+            self._read_only,
+            self._is_open,
+        )
+
+    def __setstate__(
+        self,
+        state: tuple[
+            Store,
+            int,
+            int,
+            list[str] | None,
+            dict[Any, Any],
+            dict[str | None, list[str]],
+            OrderedDict[str, Any],
+            int,
+            int,
+            bool,
+            bool,
+        ],
+    ) -> None:
+        (
+            self._store,
+            self._max_size,
+            self._current_size,
+            self._keys_cache,
+            self._contains_cache,
+            self._listdir_cache,
+            self._values_cache,
+            self.hits,
+            self.misses,
+            self._read_only,
+            self._is_open,
+        ) = state
+        self._mutex = Lock()
+
+    def __len__(self) -> int:
+        return len(self._keys())
+
+    def __iter__(self) -> Iterator[str]:
+        return self.keys()
+
+    def __contains__(self, key: str | Path) -> bool:
+        with self._mutex:
+            if key not in self._contains_cache:
+                # Check if it's a Store object vs dict-like object
+                if hasattr(self._store, "supports_listing"):
+                    # It's a Store object - use async interface
+                    try:
+                        result = asyncio.run(self._store.exists(str(key)))
+                    except RuntimeError:
+                        # Already in async context
+                        raise NotImplementedError(
+                            "Cannot use 'in' operator in async context. Use 'await store.exists(key)' instead."
+                        ) from None
+                else:
+                    # It's a dict-like object (for tests) - use sync interface
+                    result = key in cast(dict[str, Any], self._store)
+
+                self._contains_cache[key] = bool(result)
+            return bool(self._contains_cache[key])
+
+    async def clear(self) -> None:
+        """
+        Remove all keys from the store and clear the cache.
+
+        This operation clears both the underlying store and invalidates
+        all cached data to maintain consistency.
+        """
+        # Check if store is writable
+        self._check_writable()
+
+        await self._store.clear()
+        self.invalidate()
+
+    def keys(self) -> Iterator[str]:
+        """
+        Return an iterator over the keys in the store.
+        """
+        with self._mutex:
+            return iter(self._keys())
+
+    def _keys(self) -> list[str]:
+        if self._keys_cache is None:
+            # Check if it's a Store object vs dict-like object
+            if hasattr(self._store, "supports_listing"):
```

Comment:
All stores have a `supports_listing` property (you can see this defined on the base `Store` class), so this if check is redundant because it always returns `True`.